Safe panics when trying to copy from a specific version

When using a KV2 backend trying to copy from a specific version causes safe to panic.

```bash
./safe cp secret/test^3 secret/hmm
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/starkandwayne/safe/vault.(*Vault).Copy(0xc000061d00, 0x7ffeefbff4e2, 0xd, 0xc0002a3d50, 0xc, 0x0, 0x2, 0xc000155e58)
	/Users/dks0579496/go/src/github.com/starkandwayne/safe/vault/vault.go:663 +0xa2f
main.main.func32(0x13b57ec, 0x4, 0xc000384bc0, 0x2, 0x2, 0xc000372601, 0x0)
	/Users/dks0579496/go/src/github.com/starkandwayne/safe/main.go:2706 +0x2fc
main.(*Runner).Execute(0xc000060ff0, 0x13b57ec, 0x4, 0xc000384bc0, 0x2, 0x2, 0x0, 0x0)
	/Users/dks0579496/go/src/github.com/starkandwayne/safe/runner.go:117 +0x9b
main.main()
	/Users/dks0579496/go/src/github.com/starkandwayne/safe/main.go:4161 +0x2895
```

--- a/main.go
+++ b/main.go
@@ -2687,6 +2687,10 @@ and overwrite all versions of the secret at the destination.
 			}
 		}
 
+		if opt.Copy.Recurse && (vault.PathHasVersion(args[0]) || vault.PathHasVersion(args[1])) {
+			return fmt.Errorf("Cannot recursively copy a path with specific version")
+		}
+
 		//Don't try to recurse if operating on a key
 		// args[0] is the source path. args[1] is the destination path.
 		if opt.Copy.Recurse && !(vault.PathHasKey(args[0]) || vault.PathHasKey(args[1])) {

--- a/main.go
+++ b/main.go
@@ -2639,6 +2639,10 @@ and overwrite all versions of the secret at the destination.
 			}
 		}
 
+		if vault.PathHasVersion(args[1]) {
+			return fmt.Errorf("Cannot move to a specific destination version")
+		}
+
 		//Don't try to recurse if operating on a key
 		// args[0] is the source path. args[1] is the destination path.
 		if opt.Move.Recurse && !(vault.PathHasKey(args[0]) || vault.PathHasKey(args[1])) {
@@ -2687,7 +2691,11 @@ and overwrite all versions of the secret at the destination.
 			}
 		}
 
-		if opt.Copy.Recurse && (vault.PathHasVersion(args[0]) || vault.PathHasVersion(args[1])) {
+		if vault.PathHasVersion(args[1]) {
+			return fmt.Errorf("Cannot copy to a specific destination version")
+		}
+
+		if opt.Copy.Recurse && vault.PathHasVersion(args[0]) {
 			return fmt.Errorf("Cannot recursively copy a path with specific version")
 		}
 

--- a/vault/utils.go
+++ b/vault/utils.go
@@ -51,6 +51,14 @@ func PathHasKey(path string) bool {
 	return key != ""
 }
 
+// PathHasVersion returns true if the given path has a version specified in its
+// syntax.
+// False otherwise.
+func PathHasVersion(path string) bool {
+	_, _, version := ParsePath(path)
+	return version != 0
+}
+
 func Canonicalize(p string) string {
 	p = strings.TrimSuffix(p, "/")
 	p = strings.TrimPrefix(p, "/")

--- a/vault/vault.go
+++ b/vault/vault.go
@@ -670,7 +670,7 @@ func (v *Vault) Copy(oldpath, newpath string, opts MoveCopyOpts) error {
 
 		if len(t) == 0 {
 			// Prevent a panic
-			return fmt.Errorf("No valid secrets were found to be copied")
+			return NewSecretNotFoundError(srcPath)
 		}
 
 		if srcVersion != 0 {

--- a/vault/vault.go
+++ b/vault/vault.go
@@ -612,8 +612,16 @@ func (v *Vault) Copy(oldpath, newpath string, opts MoveCopyOpts) error {
 		}
 	}
 
-	srcPath, srcKey, _ := ParsePath(oldpath)
-	dstPath, dstKey, _ := ParsePath(newpath)
+	srcPath, srcKey, srcVersion := ParsePath(oldpath)
+	dstPath, dstKey, dstVersion := ParsePath(newpath)
+
+	if dstVersion != 0 {
+		return fmt.Errorf("Copying a secret to a specific destination version is not supported")
+	}
+
+	if opts.Deep && srcVersion != 0 {
+		return fmt.Errorf("Performing a deep copy of a specified version is not supported")
+	}
 
 	var toWrite []*Secret
 	if srcKey != "" { //Just a single key.
@@ -651,13 +659,28 @@ func (v *Vault) Copy(oldpath, newpath string, opts MoveCopyOpts) error {
 		t, err := v.ConstructSecrets(srcPath, TreeOpts{
 			FetchKeys:           true,
 			GetOnly:             true,
-			FetchAllVersions:    opts.Deep,
+			FetchAllVersions:    opts.Deep || srcVersion != 0,
 			GetDeletedVersions:  opts.Deep && opts.DeletedVersions,
-			AllowDeletedSecrets: opts.Deep,
+			AllowDeletedSecrets: opts.Deep || srcVersion != 0,
 		})
 
 		if err != nil {
 			return err
+		}
+
+		if len(t) == 0 {
+			// Prevent a panic
+			return fmt.Errorf("No valid secrets were found to be copied")
+		}
+
+		if srcVersion != 0 {
+			//Filter results to the specific requested secret
+			for i := range t[0].Versions {
+				if t[0].Versions[i].Number == uint(srcVersion) {
+					t[0].Versions = []SecretVersion{t[0].Versions[i]}
+					break
+				}
+			}
 		}
 
 		err = t[0].Copy(v, dstPath, TreeCopyOpts{Clear: opts.Deep, Pad: opts.Deep})

